credential/cargo-credential/src/error.rs

```diff
@@ -0,0 +1,193 @@
+use serde::{Deserialize, Serialize};
+use std::error::Error as StdError;
+use thiserror::Error as ThisError;
+
+/// Credential provider error type.
+/// 
+/// `UrlNotSupported` and `NotFound` errors both cause Cargo
+/// to attempt another provider, if one is available. The other 
+/// variants are fatal.
```

Comment:
By making it easier to use, a part of me worries that we are making it harder to correctly report the right error variant.

The way I've handled this in the error codes for my CLI is to have a sloppy error type for the implementation of each function call in `main` and then have a more constrained error type with explicit conversions used by `main`.  This does require the structure of the code to align well with the error variants.

Reply:
Thanks. I'm definitely open to feedback on improving this. 

Since it will be implemented by external crates, it's hard to know up front what all the variants should be. In the original implementation, I tried to anticipate the common ones and add variants, but in this change I decided to reduce the number of variants and have an `Other` case.

I'm not quite sure what you're suggesting here. Removing some of the `From` impls that convert to `Other` and requiring explicit conversions?

Reply:
> I'm not quite sure what you're suggesting here. Removing some of the From impls that convert to Other and requiring explicit conversions?

Not quite a suggestion but effectively yes.  Phrased differently: "I'm looking for input on whether we should reduce `From` impls to encourage explicit conversions so people are more likely to specify the correct error semantics".

Reply:
I made changes to remove the `From` impls for `serde::Error` and `io::Error` and added more error contexts.

Reply:
To be clear, I wasn't saying it had to be done but was wanting to make sure we made a conscious decision.  If this is what you feel is best great.  If not, feel free to say so.